src/common/ceph_crypto.cc

```diff
@@ -12,83 +12,122 @@
+    ceph_assert_always("openssl passed wrong mutex index" == nullptr);
   }
 
-  if (++crypto_refs == 1) {
-    NSSInitParameters init_params;
-    memset(&init_params, 0, sizeof(init_params));
-    init_params.length = sizeof(init_params);
-
-    uint32_t flags = (NSS_INIT_READONLY | NSS_INIT_PK11RELOAD);
-    if (cct->_conf->nss_db_path.empty()) {
-      flags |= (NSS_INIT_NOCERTDB | NSS_INIT_NOMODDB);
+  if (mode & CRYPTO_READ) {
+    if (mode & CRYPTO_LOCK) {
+      ssl_mutexes[mutex_num].lock_shared();
+    } else if (mode & CRYPTO_UNLOCK) {
+      ssl_mutexes[mutex_num].unlock_shared();
+    }
+  } else if (mode & CRYPTO_WRITE) {
+    if (mode & CRYPTO_LOCK) {
+      ssl_mutexes[mutex_num].lock();
+    } else if (mode & CRYPTO_UNLOCK) {
+      ssl_mutexes[mutex_num].unlock();
     }
-    crypto_context = NSS_InitContext(cct->_conf->nss_db_path.c_str(), "", "",
-                                     SECMOD_DB, &init_params, flags);
   }
-  pthread_mutex_unlock(&crypto_init_mutex);
-  ceph_assert_always(crypto_context != NULL);
 }
 
-void ceph::crypto::shutdown(bool shared)
+static unsigned long
+ssl_get_thread_id(void)
 {
-  pthread_mutex_lock(&crypto_init_mutex);
-  ceph_assert_always(crypto_refs > 0);
-  if (--crypto_refs == 0) {
-    NSS_ShutdownContext(crypto_context);
-    if (!shared) {
-      PR_Cleanup();
-    }
-    crypto_context = NULL;
-    crypto_init_pid = 0;
+  static_assert(sizeof(unsigned long) >= sizeof(pthread_t));
+  /* pthread_t may be any data type, so a simple cast to unsigned long
+   * can rise a warning/error, depending on the platform.
+   * Here memcpy is used as an anything-to-anything cast. */
+  unsigned long ret = 0;
+  pthread_t t = pthread_self();
+  memcpy(&ret, &t, sizeof(pthread_t));
+  return ret;
+}
+#endif /* not OPENSSL_API_1_1 */
+
+static void init() {
+#ifndef OPENSSL_API_1_1
+  if (++crypto_refs != 1) {
+    return;
   }
-  pthread_mutex_unlock(&crypto_init_mutex);
+
+  /* According to https://wiki.openssl.org/index.php/Library_Initialization#libcrypto_Initialization */
+  OpenSSL_add_all_algorithms();
+  ERR_load_crypto_strings();
+
+  /* Initialize locking callbacks, needed for thread safety.
+   * http://www.openssl.org/support/faq.html#PROG1
+   */
+  CRYPTO_set_locking_callback(&ssl_locking_callback);
+  CRYPTO_set_id_callback(&ssl_get_thread_id);
+
+  OPENSSL_config(nullptr);
+#endif /* OPENSSL_API_1_1 */
 }
 
-ceph::crypto::nss::HMAC::~HMAC()
-{
-  PK11_DestroyContext(ctx, PR_TRUE);
-  PK11_FreeSymKey(symkey);
-  PK11_FreeSlot(slot);
+static void shutdown() {
+#ifdef OPENSSL_API_1_1
+  if (--crypto_refs != 0) {
+    return;
+  }
+
+  /* Shutdown according to
+   * https://wiki.openssl.org/index.php/Library_Initialization#Cleanup
+   * http://stackoverflow.com/questions/29845527/how-to-properly-uninitialize-openssl
+   */
+  CRYPTO_set_locking_callback(nullptr);
+  CRYPTO_set_id_callback(nullptr);
+  ENGINE_cleanup();
+  CONF_modules_unload(1);
```

Comment:
```suggestion
  CONF_modules_free();
  CONF_modules_unload(1);
```

so we need to free the memory allocated by the configuration modules before unloading them.

Reply:
http://pulpito.ceph.com/kchai-2019-06-21_15:25:10-rados-wip-kefu-testing-2019-06-21-2037-distro-basic-smithi/4057157/